text/0000-windows-subsystem.md

```diff
@@ -0,0 +1,150 @@
+process is non-trivial:
+
+A new symbol `pub extern "system" WinMain(...)` with specific argument
+and return types must be declared, which will become the new entry point for
+the program.
+
+This is unsafe, and will skip the initialization code in `libstd`.
+
+# Detailed design
+[design]: #detailed-design
+
+When an executable is linked while compiling for a windows target, it will be
+linked for a specific *Subsystem*. The subsystem determines how the operating
+system will run the executable, and will affect the execution environment of
+the program.
+
+In practice, only two subsystems are very commonly used: `CONSOLE` and
+`WINDOWS`, and from a user's perspective, they determine whether a console will
+be automatically created when the program is started.
+
+The solution this RFC proposes is to always export both `main` and `WinMain`
+symbols from rust executables compiled for windows. The `WinMain` function
+will simply delegate to the `main` function.
+
+The exact signature is:
+```
+pub extern "system" WinMain(
+    hInstance: HINSTANCE,
+    hPrevInstance: HINSTANCE,
+    lpCmdLine: LPSTR,
+    nCmdShow: i32
+) -> i32;
+```
+
+Where `HINSTANCE` is a pointer-sized opaque handle, and `LPSTR` is a C-style
+null terminated string.
+
+All four parameters are either irrelevant or can be obtained easily through
+other means:
+- `hInstance` - Can be obtained via `GetModuleHandle`.
+- `hPrevInstance` - Is always NULL.
+- `lpCmdLine` - `libstd` already provides a function to get command line
+  arguments.
+- `nCmdShow` - Can be obtained via `GetStartupInfo`, although it's not actually
+  needed any more (the OS will automatically hide/show the first window created).
+
+The end result is that rust programs will "just work" when the subsystem is
+overridden via custom linker arguments, and does not require `rustc` to
+parse those linker arguments.
+
+A possible future extension would be to add additional command-line options to
+`rustc` (and in turn, `Cargo.toml`) to specify the subsystem directly. `rustc`
+would automatically translate this into the correct linker arguments for
+whichever linker is actually being used.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Additional platform-specific code.
+- The difficulty of manually calling the rust initialization code is potentially
+  a more general problem, and this only solves a specific (if common) case.
+- This is a breaking change for any crates which already export a `WinMain`
+  symbol. It is likely that only executable crates would export this symbol,
+  so the knock-on effect on crate dependencies should be non-existent.
+
+  A possible work-around for this is described below.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Emit either `WinMain` or `main` from `libstd` based on `cfg` options.
+
+  This has the advantage of not requiring changes to `rustc`, but is something
+  of a non-starter since it requires a version of `libstd` for each subsystem.
```

Comment:
To clarify here, the compiler generates a `main` symbol as the very last thing when an executable is generated. This main symbol is then wired up to libstd's `#[start]` function. In that sense, I don't think that this would require multiple versions of libstd?

Also, are you thinking that this alternative is basically `rustc --cfg subsystem=windows foo.rs` or something like that?